### ISSUE TITLE
Unique id for all weapons ([attack])

### ISSUE THREAD

**Issue by ProditorMagnus** [wesnoth/wesnoth#9324]
### Describe the desired feature

There should be way to identify weapon of unit by better way than [attack] index. Currently weapons have name, but one unit is able to have multiple weapons with same name.

So far we have discussed 2 options.
1) Add new id key to weapon. When id is not provided, engine generates it. Id is required to be unique within weapons of this unit, and nice to have unique globally. If user is allowed to manually specify id and manually specifies conflicting id during attack creation, attack creation should fail.
2) Change existing name key to act as id. This means making additional attacks of this unit with same name be dropped (same semantics as unit id conflict).

I support option 1. @CelticMinstrel supports option 2.

**Comment by ProditorMagnus** [comment]
Example of violation in mainline https://github.com/wesnoth/wesnoth/blob/master/data/core/units/dunefolk/Blademaster.cfg#L28-L37

**Comment by CelticMinstrel** [comment]
Just to be clear, it only needs to be unique per unit, not globally unique.

**Comment by rodolphecombe** [comment]
Or you can filter by me surviving faire it the attack? 
Like name+range+damage_type? 
I support option 1 as option 2 will break a lot of UMC that use current attack names

**Comment by gfgtdf** [comment]
I'm trying to thinkof advantages of the options

Advantages of  (1)/Disadvantages of (2)
- Id is imo the better name for what it does
- We can enforce the "ids have to be unique" rule in the c++ code wihout breaking umc.
- We can also change exiting names when using a new key, in particular to remove spaces from it.
- Id can probably be generated from name= if not given

Advantages of  (2)/Disadvantages of (1)
- Having both Id and name is confusing, and if we want to end up derecating  using name= in the long run it will probably annoy umc devs even more than with options (2)
- Generating the id in a smart way might turn out to be complicated

i think so far i prefer option 1 more but im not so sure

**Comment by ProditorMagnus** [comment]
> Advantages of (2)/Disadvantages of (1)
> 
> * Having both Id and name is confusing, and if we want to end up derecating  using name= in the long run it will probably annoy umc devs even more than with options (2)
> * Generating the id in a smart way might turn out to be complicated

I think it is reasonable to keep name in addition to id

1. name specifies icon
2. name can refer to group of attacks, for animation filtering

For id generation some ideas

1. `(unit id) + ("-attack-") + (unit-local index)` - locally unique attack index, together with unit id making globally unique id. When weapon is removed from unit and new is added, previous weapon id should not be reused.
2. `("attack-") + (new value from id sequence, like for unit id)` - globally unique attack id from sequence.
3. There may be different generated id pattern between [unit_type] weapons, and weapons given later. For [unit_type] weapons could have `(unit type id) + ("-attack-") + (attack index)`.

**Comment by ForestDragon-wesnoth** [comment]
Option 2 would be extremely awful for UMC, knowing from personal experience how many add-ons often use identical names for multiple attacks.

Option 1 seems alright. But if that doesn't work, then I would rather keep the current system than have the widespread heavy damage to UMC that would be caused by option 2.

**Comment by hladkyjakub** [comment]
Can't we just leave it "name" to break nothing?

**Comment by DuncanusDillius** [comment]
I agree with FD here, Option 1 seems good but i would rather keep the same and not break UMC

**Comment by Roge-Tebnelok** [comment]
Agreed, it's probably better to leave attack filtering to just "name". Also, units that have multiple attacks with same name are quite rare in mainline, not quite enough to make such a change?

**Comment by hladkyjakub** [comment]
> Agreed, it's probably better to leave attack filtering to just "name". Also, units that have multiple attacks with same name are quite rare in mainline, not quite enough to make such a change?

Definitely enought to make name act as id (it doesn't change anything for players)

**Comment by ProditorMagnus** [comment]
If this issue option 1 is implemented, then #9670 is not needed. However, if there is no agreed solution for 9324, then 9670 will at least fix mainline so that anyone who wants to have unique weapons for their addon is able to do it without having to recreate mainline unit types.

Note that option 1 will keep support to filter by name. Filtering by id will be allowed in addition.

Looks like there is agreement that option 2 should not be done.

**Comment by CelticMinstrel** [comment]
> 1. name specifies icon
> 2. name can refer to group of attacks, for animation filtering

1. We already have icon for that, no point having two keys that do the same thing.
2. People use it for that, but it doesn't really serve that purpose well in general. It would be better to have something like `tags` for filtering purposes.

> Also, units that have multiple attacks with same name are quite rare in mainline, not quite enough to make such a change.

Nope, as you can see from the linked PR, there are quite a few of them.

> RPG mods rely on the name attribute to classify weapons, as well as some others.

But the problem is, it doesn't work for classifying weapons. If you want "all swords", for example, `name=sword` won't match them all. You also need `name=saber` and maybe a few others. Wouldn't it be easier if you can just write `tags=sword` and it'll match every sword? (Though obviously that's contingent on people using the tag system when they define the weapon; but we'd update all mainline ones and make `name` be treated as a tag if present, for backwards compatibility.)

---

I'm no longer in favour of option 2 either. We should deprecate `name` (but not remove it) and add `id` and `tags`.

So, we'll have 4 keys in total:

* `id` - used to uniquely identify an attack on a unit. If not specified, but `name` is, it'll default to `name`, possibly with a digit appended to maintain uniqueness.
* `tags` - a comma-separated list of tags that can be used for filtering. For example, a saber could have `tags=sword,saber` so that it will match a filter for `tags=sword`. If `name` is present, it'll be appended to the tags list.
* `icon` - as now, the path to the icon. If not specified, but `name` is, that'll be used to select a default icon. If neither is specified, probably use a blank square.
* `name` - deprecated, but will still continue to function in the way it always has for the foreseeable future.

**Comment by ProditorMagnus** [comment]
This proposal looks good for me. I would prefer deprecation level in a way that before 1.21 it will not be visible in chat area.

**Comment by ForestDragon-wesnoth** [comment]
Deprecating name= means it will be tempting for developers to entirely remove the feature down the line, even if it takes multiple release cycles.

To quote discussion taking place on the official discord:

dwarftough: "deprecate often means subsequent removing? Even if not, it may cause troubles like all old content will elicit a warning for all units using name, even if no filtering done
like I don't see as tricky or broken if you just define a unit"

celticminstrel [IRC]: "Technically, yes. But that subsequent removal will be several release cycles down the line, if ever."

Pentarctagon: "it's essentially developer CYA
we're not saying it will be removed
but if it is, well we already told you it could be"

vultraz: "Never assume deprecated code will remain in place when I’m around"




In general, the more developers break backwards compatibility for WML code, the harder it becomes to port mods (especially ones released multiple stable versions ago, where code differences start to pile up), and the less mods there are on each new stable versions.

On top of that, flooding UMC creators with deprecation messages over entirely normal functioning attacks will make them more likely to turn off deprecation messages in settings and risk missing actually important deprecation messages.

**Comment by CelticMinstrel** [comment]
> Deprecating name= means it will be tempting for developers to entirely remove the feature down the line, even if it takes multiple release cycles.

Uhh, yes, maybe removing it at some point in the future is the entire point of deprecation. But we would deprecate it at level 1 most likely.


>In general, the more developers break backwards compatibility for WML code, the harder it becomes to port mods (especially ones released multiple stable versions ago, where code differences start to pile up), and the less mods there are on each new stable versions.

Well, this is why we try _not_ to remove the deprecated things as much as possible. That way you don't need to update your deprecated code just to release your add-on on the new version – you can just copy it over, playtest a bit to make sure it's all working, and release it. Then, update deprecated code bit by bit as you please.

> On top of that, flooding UMC creators with deprecation messages over entirely normal functioning attacks will make them more likely to turn off deprecation messages in settings and risk missing actually important deprecation messages.

It's supposed to be perfectly fine to ignore deprecation warnings, so if they want to ignore them then that's their choice. It's not a great idea to do so, admittedly. But the point of deprecation is so that you don't need to update your code right away. You can do it whenever.

> It's not like deprecating/replacing a niche feature that barely gets used anywhere, this is deprecating something that gets used in literally every single add-on with a custom unit.

Avoiding a deprecation on the basis of how widely used a feature is sounds like the most ridiculous argument ever. Avoiding _removal_ on that basis is reasonable, but deprecation is just adding a message like "please don't use this for new code". It does not (if done right – there have been some mistakes in the past) affect the functioning of anyone's existing code.

**Comment by Refumee** [comment]
So, I’ve read everything thoroughly.

If I understand correctly, the issue is that the naming of attacks for the core units is unsatisfying?

The proposed solution is to add an id and a tag, while deprecating name. Deprecating means that name will eventually be removed in the future. That’s how I interpret it.

Why isn’t the solution simply to rename the attacks of the core units?
Wouldn’t that require much less effort?
I’m not sure how many UMCs this would affect, but I imagine it wouldn’t be that many.

My suggested solution:
Keep name and introduce tag as an additional identifier for the attacks. Then, rework the attack names for the core units. As mentioned, this would require additional checks.

**Comment by hladkyjakub** [comment]
Whats bad on improving name functionality (changing names of mainline duplicate weapons) and adding id's and tags for pproper treatment too?

**Comment by ProditorMagnus** [comment]
When id generation works, then mainline names do not need current PR change. My plan is that some point during January I check status and

If 9324 is open
- ask for merge of 9670
- add reminder comment to revert 9670 after 9324 is implemented

If 9324 is fixed
- close or revert 9670

Else
- ask for merge of 9670

**Comment by Discontinuum** [comment]
> But the problem is, it doesn't work for classifying weapons. If you want "all swords", for example, name=sword won't match them all. You also need name=saber and maybe a few others. Wouldn't it be easier if you can just write tags=sword and it'll match every sword? (Though obviously that's contingent on people using the tag system when they define the weapon; but we'd update all mainline ones and make name be treated as a tag if present, for backwards compatibility.)

Well, people managed to overcome these conditions by making big lists, like in LotI there is a list that swords are `sword`, `saber`, `greatsword` and some more. It may be improved, sure, but breaking existing names overnight for it doesn't sound good to me. My key concern is that old names should stay for now.

Adding new keys, `id=` and `tags=`, is good, that seems a good way to me. A couple of questions though, what is the primary benefits of `tags` succeeding `name`? A possibility to assign multiple tags and not changing the old names, or something more? Also, is there a risk of `tags=` becoming new `name` in terms of confusion?

**Comment by rodolphecombe** [comment]
> It's supposed to be perfectly fine to ignore deprecation warnings

Except that it floods the log

> Why isn’t the solution simply to rename the attacks of the core units?
Wouldn’t that require much less effort?
I’m not sure how many UMCs this would affect, but I imagine it wouldn’t be that many.

This is the PR. It means renaming quite a few attacks and that will break many addons.

**Comment by hladkyjakub** [comment]
> This is the PR. It means renaming quite a few attacks and that will break many addons. 
> 
Which ones?
I don't know any addon that would reffer a weapon of a unit with two same named weapons by name.

**Comment by AnekronCZ** [comment]
I'm fairly neutral about this issue, but I feel like the discussion mostly focuses on the negatives. I'd like to mention that having weapon ids that are consistent could be a boon to add-on makers as well.

When there's a flaming sword on the map, for example, you could just code that it can be picked up by anyone with a "tag=sword" attack. The attack itself could be named whatever (greatsword, shortsword, Anduril, who cares) and the item would be therefore compatible with any add-ons adding / modifying units. I for one would definitely use this, as opposed to the more cumbersome methods I need to employ now.

I've coded my first add-on back in 1.10 days. I'm very grateful that I now have every QoL comforts of 1.18 at my disposal, since some stuff I can code easily in WML nowadays would require lua back when I started. I'm sure the changes broke many add-ons along the way (including mine between 1.16 and 1.18), but I still consider them a net positive.

**Comment by MrTitainin** [comment]
The mods I created or maintain would probably need some reworks, but I wholeheartedly support option 1 anyway, as I keep hitting the issue with lack of attack id.

I'd also love to see tags= key as well.

But I also think it'd be the best to set/keep name as basis for default id, tags and icon, so that new code will seamlessly blend with old add-ons, especially those that already implement name uniqueness.

**Comment by rodolphecombe** [comment]
> It's supposed to be perfectly fine to ignore deprecation warnings

Except that it floods the log

> Why isn’t the solution simply to rename the attacks of the core units?
Wouldn’t that require much less effort?
I’m not sure how many UMCs this would affect, but I imagine it wouldn’t be that many.

This is the PR. It means renaming quite a few attacks and that will break many addons. 

> Which ones? I don't know any addon that would reffer a weapon of a unit with two same named weapons by name.

At least Loti and my two add-ons. Probably sea fire too

**Comment by ProditorMagnus** [comment]
Perhaps tags is not needed, instead name can support comma separated values (first of which is used to generate id). That avoids one issue of deprecation as well - basically even if we deprecate name, mainline can never remove it from core attacks since that would break the code that expects meaningful name value.

**Comment by ForestDragon-wesnoth** [comment]
> Perhaps tags is not needed, instead name can support comma separated values (first of which is used to generate id). That avoids one issue of deprecation as well - basically even if we deprecate name, mainline can never remove it from core attacks since that would break the code that expects meaningful name value.

That sounds like a good idea. For single-name attacks the code/interactions are unchanged, but modders can intentionally add more values to use the name field as tags

**Comment by kameramen0106** [comment]
> Perhaps tags is not needed, instead name can support comma separated values (first of which is used to generate id). That avoids one issue of deprecation as well - basically even if we deprecate name, mainline can never remove it from core attacks since that would break the code that expects meaningful name value.

Oh, this is a stylish idea and might solve this problem nicely. I agree with your idea.

**Comment by rodolphecombe** [comment]
> Perhaps tags is not needed, instead name can support comma separated values (first of which is used to generate id). That avoids one issue of deprecation as well - basically even if we deprecate name, mainline can never remove it from core attacks since that would break the code that expects meaningful name value.

Do you think you can implement it?

**Comment by rodolphecombe** [comment]
> If you're worried about flooding a log that you don't even see unless you go hunting for it, then the problem is you.

??? I don't even know how to answer such an absurd comment. So you mean that was a dev your never use the log?

**Comment by CelticMinstrel** [comment]
> I don't even know how to answer such an absurd comment. So you mean that was a dev your never use the log?

As an _add-on_ developer I almost never have a reason to look at the log files, yes.

The log file is indispensable when debugging the game itself, but for an add-on author it's not all that useful. It's not like it's completely useless – there are some things (most notably Wesnoth Formula Language errors) that you can only see in the log – but its use is quite limited.

Personally, I like having deprecation messages thrown in my face in the chat log, because I'd prefer to fix them sooner rather than later. But that viewpoint seems to be in the minority, so I think it would be worth removing those, maybe with an advanced preference to re-enable them. Many deprecation messages are now placed in the Lua console nowadays anyway, so I wouldn't even need to enable such a setting very often.

And a _player_ should pretty much _never_ be looking at the log file unless they're reporting a bug. So reducing the size of the log file for the sake of players is pure nonsense, and reducing the size of the log file for add-on developers is highly questionable as well.

**Comment by gfgtdf** [comment]
> The log file is indispensable when debugging the game itself, but for an add-on author it's not all that useful. It's not like it's completely useless – there are some things (most notably Wesnoth Formula Language errors) that you can only see in the log – but its use is quite limited.

This is not true, a lot of the actual bugs in umc add-ons do show as warnings in the log and give you  a big hint on whats going wrong, and also the other way around an warning in he log is quite often an error or at least a bad code.

I say this as someone who often tried to help people with their umc questions/erros,  just to find out in the majority of cases that the answer is right there  as a warning or even an error in the log, and that people asked me without even bothering to look there first and somehow didn't see a problem with multiple warnings in their logs. At some point didn't even bother to help people anymore unless their addon gave an empty log 

We should  definitely not teach people to ignore the log warnings.

**Comment by gfgtdf** [comment]
> > Advantages of (2)/Disadvantages of (1)
> > 
> > * Having both Id and name is confusing, and if we want to end up derecating  using name= in the long run it will probably annoy umc devs even more than with options (2)
> > * Generating the id in a smart way might turn out to be complicated
> 
> I think it is reasonable to keep name in addition to id

Agreed.

> For id generation some ideas
> 
>     1. `(unit id) + ("-attack-") + (unit-local index)` - locally unique attack index, together with unit id making globally unique id. When weapon is removed from unit and new is added, previous weapon id should not be reused.
> 
>     2. `("attack-") + (new value from id sequence, like for unit id)` - globally unique attack id from sequence.
> 
>     3. There may be different generated id pattern between [unit_type] weapons, and weapons given later. For [unit_type] weapons could have `(unit type id) + ("-attack-") + (attack index)`.

I think when a weapon is removed and then re-added we usually want the same id to be used, in particular since rebuildng a unit works by resetting it and then re-adding all objects, we don't want a rebuild of a unit to change the id of a weapon that was added by an object.

A even more complicated situation arises when we have two weapons added via different objects: it would be nice if removing the first weapon via `[remove_object]/u:remove_modification()` would not change the id of the second object (note that remove_object also works by rebuilding the unit)

**Comment by Discontinuum** [comment]
> If you're worried about flooding a log that you don't even see unless you go hunting for it, then the problem is you.

Well, firstly, saying things like "the problem is you" is borderline insulting and doesn't facilitate the discussion.

Secondly, as gfgtdf noted, the log is very useful for add-on development and add-on debugging, especially if it's some kind of a complicated add-on, like World Conquest and forks or Legend of the Invincibles. I run Wesnoth from the command line, I remember in 1.16 the log was printed in stdout, and now it's just in files, and I didn't quite like it. 

> Big lists is a workaround, however. And, it can never catch every imaginable case. As soon as someone decides to name their sword attack something exotic like "waveblade" your big list fails.

I agree that it's a workaround, doesn't catch all imaginable cases and you have to add new ones as you stumble on them. But usually you have a pool of units that you work with, and to make it work for most people you just cover all cases for all supposed units. 

And I have to note that the other way doesn't catch all imaginable cases either. For the `tags=` system to work, you need some kind of convention being maintained among the add-on unit creators about tagging the attacks. Even following the convention wouldn't prevent cases like "my add-on disagrees with the unit creator about tags". I don't deny that tags can make handle this complexity better but they don't provide universality.

**Comment by Discontinuum** [comment]
Regarding a proposal to make `name` accept comma-separated values, I think it still leads to an API break: both from unexpected attacks potentially being filtered and from adoption by mainline units (they obviously will be the first ones to adopt the new `name`, I believe, so `name`s will change for them), so I agree with CelticMinstrel's stance on that.

> This approach does mean that old code that filtered on name=sword will suddenly start to match attacks with name=saber,sword which they previously didn't. This is also an API break and may be an argument against this approach.

**Comment by rodolphecombe** [comment]
> This approach does mean that old code that filtered on name=sword will suddenly start to match attacks with name=saber,sword which they previously didn't. This is also an API break and may be an argument against this approach.

That is indeed a good point. However I want to note that the original proposal was too do the opposite: old code that filtered on name=sword would suddenly start to not match some attacks that were matching before

**Comment by CelticMinstrel** [comment]
> This is not true, a lot of the actual bugs in umc add-ons do show as warnings in the log and give you a big hint on whats going wrong, and also the other way around an warning in he log is quite often an error or at least a bad code.
> 
> I say this as someone who often tried to help people with their umc questions/erros, just to find out in the majority of cases that the answer is right there as a warning or even an error in the log, and that people asked me without even bothering to look there first and somehow didn't see a problem with multiple warnings in their logs. At some point didn't even bother to help people anymore unless their addon gave an empty log

Alright, fair point. It is certainly worth looking at the log file if you encounter an issue that you can't figure out. But it isn't really something worth watching while you play.

> And I have to note that the other way doesn't catch all imaginable cases either. For the tags= system to work, you need some kind of convention being maintained among the add-on unit creators about tagging the attacks. Even following the convention wouldn't prevent cases like "my add-on disagrees with the unit creator about tags". I don't deny that tags can make handle this complexity better but they don't provide universality.

True enough, but it comes _closer_ to universality than maintaining a big list.

> That is indeed a good point. However I want to note that the original proposal was to do the opposite: old code that filtered on name=sword would suddenly start to not match some attacks that were matching before

What do you mean by "the original proposal"? The _current_ proposal (deprecate name in favor of id/tags/icon) doesn't have this property.

**Comment by rodolphecombe** [comment]
Sorry, I meant option 2 of the proposal, that is reflected in PR #9670
